runtimevar/drivertest/drivertest.go

```diff
@@ -392,3 +409,92 @@ func testDelete(t *testing.T, newHarness HarnessMaker) {
 		t.Errorf("got UpdateTime %v < previous %v, want >=", state.UpdateTime(), prev.UpdateTime())
 	}
 }
+
+func testUpdateWithErrors(t *testing.T, newHarness HarnessMaker) {
+	const (
+		name     = "test-updating-variable-to-error"
+		content1 = `[{"Name": "Foo", "Text": "Bar"}]`
+		content2 = "invalid-json"
+		content3 = "invalid-json2"
+	)
+	want := []*Message{{Name: "Foo", Text: "Bar"}}
+
+	h, err := newHarness(t)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer h.Close()
+	if !h.Mutable() {
+		return
+	}
+	ctx := context.Background()
+
+	// Create the variable and verify WatchVariable sees the value.
+	if err := h.CreateVariable(ctx, name, []byte(content1)); err != nil {
+		t.Fatal(err)
+	}
+
+	var jsonData []*Message
+	drv, err := h.MakeWatcher(ctx, name, runtimevar.NewDecoder(jsonData, runtimevar.JSONDecode))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		if err := drv.Close(); err != nil {
+			t.Error(err)
+		}
+	}()
+	state, _ := drv.WatchVariable(ctx, nil)
+	if state == nil {
+		t.Fatal("got nil state, want a non-nil state with a value")
+	}
+	got, err := state.Value()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if gotSlice, ok := got.([]*Message); !ok {
+		t.Fatalf("got value of type %T expected []*Message", got)
+	} else if !cmp.Equal(gotSlice, want) {
+		t.Errorf("got %v want %v", gotSlice, want)
+	}
+
+	// Update the variable to invalid JSON and verify WatchVariable returns an error.
+	if err := h.UpdateVariable(ctx, name, []byte(content2)); err != nil {
+		t.Fatal(err)
+	}
+	state, _ = drv.WatchVariable(ctx, state)
+	if state == nil {
+		t.Fatal("got nil state, want a non-nil state with an error")
+	}
+	got, err = state.Value()
+	if err == nil {
+		t.Fatal("got nil want invalid JSON error")
+	}
+
+	// Update the variable again, with different invalid JSON.
+	// WatchVariable should block or return nil since it's the same error as before.
+	if err := h.UpdateVariable(ctx, name, []byte(content3)); err != nil {
+		t.Fatal(err)
+	}
+	tCtx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+	defer cancel()
+	state, _ = drv.WatchVariable(tCtx, state)
+	if state == nil {
+		// OK: nil indicates no change.
+	} else {
+		// WatchVariable should have blocked until tCtx was cancelled, and we
+		// should have gotten that error back.
+		got, err := state.Value()
+		if err == nil {
+			t.Fatalf("got %v want error", got)
```

Comment:
s/error/nil

Reply:
No, this is correct -- the test is expecting that err should not be nil, and it was nil. I clarified it a bit.